File: app/javascript/channels/chat_room_channel.js
import consumer from "./consumer"

function insertMessage(data, current_user_id){
  var message = data.resource;
  var class_name = current_user_id === message.user_id ? 'message-sent' : 'message-received';
  var li_side = current_user_id === message.user_id ? 'float: right;' : '';
  $('#messages').append(`<li class="col-12" style="${li_side}">
        <div>
          <div class="${class_name}">
            ${message.body}
          </div>
        </div>
      </li>`);

  // scroll to bottom of the chat
  var objDiv = document.getElementById("chat");
  objDiv.scrollTop = objDiv.scrollHeight;
}

function SendMessage(){
  $(document).on('submit', '.new_message', function(e) {
    e.preventDefault();
    var values = $(this).serializeArray();
    var data = { };
    $.each(values, function (index, value) {
      data[value.name] = value.value;
    })
    consumer.conversation.speak(data);
    $(this).trigger('reset');
  });
}

$(function() {
  consumer.conversation = consumer.subscriptions.create({ channel: "ChatRoomChannel",
    chat_room_id: $("#messages").data('chat-room-id')} , {
    connected: function() {},

    disconnected: function() {},

    received: function(data) { insertMessage(data, current_user_id) },

    speak: function(data) {
      return this.perform('speak', data);
    }
  });
  SendMessage();
});

File: app/channels/chat_room_channel.rb
class ChatRoomChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_room/#{params['chat_room_id']}" if params['chat_room_id'].present?
  end

  def unsubscribed
    stop_all_streams
  end

  def speak(data)
    return unless chat_member?(data)

    Mutators::ChatMessages::Create.call(chat_message_params(data))
  end

  private

  def chat_member?(data)
    ChatRoom.find(data['chat_room_id']).member_ids.include?(data['user_id'].to_i)
  end

  def chat_message_params(data)
    data.slice('chat_room_id', 'user_id', 'body')
  end
end
